tools/benchmark/src/main/java/com/datadog/benchmark/internal/reader/MemoryVitalReader.kt

```diff
@@ -7,16 +7,41 @@
 package com.datadog.benchmark.internal.reader
 
 import java.io.File
+import java.util.concurrent.Executors
+import java.util.concurrent.ScheduledExecutorService
+import java.util.concurrent.TimeUnit
 
 /**
  * Reads the device's `VmRSS` based on the `/proc/self/status` file.
  * cf. documentation https://man7.org/linux/man-pages/man5/procfs.5.html
  */
 internal class MemoryVitalReader(
-    internal val statusFile: File = STATUS_FILE
+    internal val statusFile: File = STATUS_FILE,
+    private val readerDelay: Long = STATUS_FILE_READ_INTERVAL_MS
 ) : VitalReader {
 
-    override fun readVitalData(): Double? {
+    private var maxMemory: Double = 0.0
+    private val executorService: ScheduledExecutorService = Executors.newSingleThreadScheduledExecutor()
+
+    private val readRunnable: Runnable = Runnable {
+        maxMemory = (readStatusFile() ?: 0.0).coerceAtLeast(maxMemory)
+    }
+
+    override fun readVitalData(): Double {
```

Comment:
can you please explain this change? shouldn't it be external caller which is making the call to `readVitalData` control the frequency of the reads?

Reply:
Actually the external caller does have a settled frequency (10s by default), but to keep this external caller simple and agnostic about all types of reader, it just collects data from all the readers and export the data immediately every 10s. So every reader has to prepare their own data to be read if they have its own logic (like here memory needs to read more frequently and only expose the maximum data during every read, but CPU reader doesn't have to).

Reply:
> external caller does have a settled frequency (10s by default)

This was an approach designed for the SDK, where all the readers were queried at the same frequency. Now, if we are designing the solution for our internal needs and can allow ourselves more frequent querying, maybe it is worth to think of just of independent frequency for all the readers instead?

Reply:
You mean that instead of having an internal frequency of every reader, it should be external caller who decides the different frequency of each reader?
I think with the solution of internal frequency of every reader, the external caller can be agnostic, and easier to test each reader independently.

Reply:
I mean right now with such solution you have 2 frequencies - of external caller and internal of the reader, which requires you to have different threads and introduce thread-safety guards. It is not quite clear why should we have 2 different frequencies while we can have only one - external, which still can be defined individually for each reader while having a single thread.

Reply:
Here the benefits that I see to separate internal frequency and external frequency is that:
*  the external caller frequency is the one that can be configured in the benchmark app, it can control both query interval of all readers, and it is also the frequency of upload. It depends on how frequently we want to sample the data.
* Having an internal frequency which can't be changed from outside can make sure that we have always the accurate vital data from each reader. external caller or app doesn't have to care about the frequency of detailed vital and doesn't have to store a cache like `maxMemory` during the time for example.

What do you think about it?